### PR TITLE
Sketch out a codec for coarse quantization (IVFFlat).

### PR DESCRIPTION
**Pull request by jtibshirani**
**Note:** this PR is rough and is not meant for detailed review. If we decide we'd like to go this direction, I will put together a proper PR and feature branch.

This PR sketches out a kNN approach based on coarse quantization (IVFFlat). It introduces a new format `VectorsFormat`, which simply delegates to `DocValuesFormat` and `PostingsFormat` under the hood:
* The original vectors are stored as `BinaryDocValues`.
* The vectors are also clustered, and the cluster information is stored in postings format. In particular, each cluster centroid is encoded to a `BytesRef` to represent a term. Each document belonging to the centroid is added to the postings list for that term.

The main goal of the sketch is to give a structure to build off when experimenting with k-means algorithms and other improvements. Currently only ‘coarse quantization’ is applied, but the format could be extended to incorporate product quantization as well (which @mayya-sharipova has been exploring in https://github.com/elastic/elasticsearch/pull/51243).

There are some pretty big hacks and open API questions. A few notes:
* Instead of adding `VectorsFormat`, I think we could just add a new docvalues type like `VectorDocValues`. The current approach seemed a bit cleaner and more self-contained (and also gives more flexibility as we experiment with the codec).
* Instead of exposing all internal data structures through `VectorValues`, we could imagine hiding most implementation details within the vectors format. I opted not to do this in the initial sketch to keep it simple.
* In order to be able to write postings, I introduced a hacky inner class `ClusterFields` that presents the clusters as if they were terms information.

I’m working on running an initial round of benchmarks and will post the results here.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/index/VectorClusteringMethod.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    ClusterCentroids(float[][] centroids) {
+      this.centroids = centroids;
+
+      this.encodedCentroids = new BytesRef[centroids.length];
+      for (int c = 0; c < centroids.length; c++) {
+        this.encodedCentroids[c] = VectorField.encode(centroids[c]);
+      }
+    }
+
+    public BytesRef findClosestCentroid(BinaryDocValues values) throws IOException {
```
Comment (mayya-sharipova) [inline lucene/core/src/java/org/apache/lucene/index/VectorClusteringMethod.java:153]: I assume we also want to have a method `findKClosestCentroids(BinaryDocValues values, int k)` for k-probes search or do we want to keep it specific to the query?

File: lucene/core/src/java/org/apache/lucene/document/FieldType.java [hunk 2]
```diff
@@ -351,6 +353,24 @@ public int pointNumBytes() {
...
+    }
+
+    if (numDimensions > VectorField.MAX_DIMS) {
+      throw new IllegalArgumentException("vector numDimensions must be <=" + VectorField.MAX_DIMS
+          + " dimensions; got " + numDimensions);
+    }
+
+    this.vectorDimensions = numDimensions;
+  }
+
```
Comment (mayya-sharipova) [inline lucene/core/src/java/org/apache/lucene/document/FieldType.java:373]: I assume later we want to put `vectorDimensions` into `hashCode` and `equals` functions as well

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/VectorClusterBackedFields.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+import org.apache.lucene.index.BaseTermsEnum;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.Fields;
+import org.apache.lucene.index.ImpactsEnum;
+import org.apache.lucene.index.PostingsEnum;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.util.BytesRef;
+
+public class VectorClusterBackedFields extends Fields {
```
Comment (mayya-sharipova) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/VectorClusterBackedFields.java]: Doesn't  `Lucene90VectorsWriter` contains the same code?

File: lucene/core/src/java/org/apache/lucene/search/VectorDistanceQuery.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+  private final String field;
+  private final float[] queryVector;
+  private final int numProbes;
+
+  public VectorDistanceQuery(String field, float[] queryVector, int numProbes) {
+    this.field = field;
+    this.queryVector = queryVector;
+    this.numProbes = numProbes;
+    if (this.numProbes <= 1) {
+      throw new IllegalArgumentException("The numProbes parameter must be > 1.");
```
Comment (mayya-sharipova) [inline lucene/core/src/java/org/apache/lucene/search/VectorDistanceQuery.java:50]: "must be > 1." => "must be >=1"

File: lucene/core/src/java/org/apache/lucene/search/ExactVectorDistanceQuery.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+  private static class ExactVectorDistanceWeight extends Weight {
+    private final ScoreMode scoreMode;
+    private final float boost;
+
+    private final String field;
+    private final float[] queryVector;
+
+    ExactVectorDistanceWeight(Query query, float boost, ScoreMode scoreMode,
+                              String field, float[] queryVector) {
```
Comment (mayya-sharipova) [inline lucene/core/src/java/org/apache/lucene/search/ExactVectorDistanceQuery.java:90]: Looks like `scoreMode` is not necessary here.

File: lucene/core/src/java/org/apache/lucene/search/ExactVectorDistanceQuery.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+        @Override
+        public float getMaxScore(int upTo) {
+          return boost;
+        }
+
+        @Override
+        public float score() throws IOException {
+          BytesRef encodedVector = docValues.binaryValue();
+          double dist = VectorField.l2norm(queryVector, encodedVector);
```
Comment (mayya-sharipova) [inline lucene/core/src/java/org/apache/lucene/search/ExactVectorDistanceQuery.java:123]: may be we can have an extra parameter for the query –– what kind of distance to calculate (cosineSimilarity, dotProduct..). But this is not necessary for initial stage

File: lucene/core/src/java/org/apache/lucene/search/VectorDistanceQuery.java [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+  }
+
+  @Override
+  public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+    return new VectorDistanceWeight(this, boost, scoreMode, field, queryVector, numProbes);
+  }
+
+  @Override
+  public String toString(String field) {
+    return String.format(Locale.ROOT, "VectorDistanceQuery{field=%s;fromQuery=%s;numCentroids=%d}",
```
Comment (mayya-sharipova) [inline lucene/core/src/java/org/apache/lucene/search/VectorDistanceQuery.java:61]: numCentroids => numProbes?

File: lucene/core/src/java/org/apache/lucene/search/VectorDistanceQuery.java [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+import org.apache.lucene.codecs.VectorValues;
+import org.apache.lucene.document.VectorField;
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.search.similarities.BooleanSimilarity;
+import org.apache.lucene.search.similarities.Similarity;
+import org.apache.lucene.util.BytesRef;
+
+public class VectorDistanceQuery extends Query {
```
Comment (mayya-sharipova) [inline lucene/core/src/java/org/apache/lucene/search/VectorDistanceQuery.java:39]: may be rename this query as ClusteringVectorDistanceQuery or something like this?

I was thinking about a structure where we have  a base class `VectorDistanceQuery`  that takes queryVector, docsIdSetIterator and  a distanceFunction (some kind of Enum that we can add later) and can score docs base on this distance function.  
`ExactVectorDistanceQuery` can extend this base class and provide  docsIdSetIterator as an full docValues.
This clustering vector query can extend this base class and provide docsIdSetIterator as a union of all docsIds from probes.

I don't have much experience with scorers, but was thinking this organization could be faster than DisjunctionScorer

File: lucene/core/src/java/org/apache/lucene/search/VectorDistanceQuery.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+      PriorityQueue<Map.Entry<BytesRef, Double>> queue = new PriorityQueue<>(
+          (first, second) -> -1 * Double.compare(first.getValue(), second.getValue()));
+
+      while (true) {
+        BytesRef encodedCentroid = centroids.next();
+        if (encodedCentroid == null) {
+          break;
+        }
+
+        double dist = VectorField.l2norm(queryVector, encodedCentroid);
```
Comment (mayya-sharipova) [inline lucene/core/src/java/org/apache/lucene/search/VectorDistanceQuery.java:166]: do we also consider to have a function `l2normsquared` that can be used for cases where we don't care about scores, just ranks. For example for this case.

**Review by mayya-sharipova — COMMENTED** [review]
Thanks @jtibshirani, the PR looks very good to me. I have done the initial review but haven't looked at all files yet.

**Comment by jtibshirani** [comment]
Thanks @mayya-sharipova for the comments. The PR is quite rough and isn't ready for a detailed review -- I was hoping to get feedback on the overall structure of the approach.

**Comment by jtibshirani** [comment]
Initial round of benchmarks on `glove-100-angular`, k=10:
```
APPROACH                          RECALL     QPS
LuceneExact()                     1.000        6.590
LuceneCluster(n_probes=5)         0.680      192.743
LuceneCluster(n_probes=10)        0.766      146.746
LuceneCluster(n_probes=20)        0.835       98.448
LuceneCluster(n_probes=50)        0.905       50.033
LuceneCluster(n_probes=100)       0.948       27.855
```

It looks like we're not achieving the same speed-up we saw in our ES coarse quantization prototype, there are likely some search inefficiencies that I need to dig into.

**Comment by jtibshirani** [comment]
Initial round of benchmarks on `glove-100-angular`, k=10:
```
APPROACH                          RECALL     QPS          DIST COMPS
LuceneExact()                     1.000        6.590      1183514
LuceneCluster(n_probes=5)         0.680      216.592         6876
LuceneCluster(n_probes=10)        0.766      167.320        12456
LuceneCluster(n_probes=20)        0.835      114.232        23286
LuceneCluster(n_probes=50)        0.905       59.715        54539
LuceneCluster(n_probes=100)       0.948       33.578       105303
```

It looks like we're not achieving the same speed-up we saw in our ES coarse quantization prototype, there are likely some search inefficiencies that I need to dig into.

**Comment by jtibshirani** [comment]
After some digging, it turns out that the benchmarking framework adds ~3ms of latency (due to the overhead of calling from python into java from py4j). I measured the 3ms overhead by replacing the vector distance query with a `MatchNoDocsQuery` and looking at the QPS.

If we subtract out the overhead, the results look roughly as follows:

```
APPROACH                          RECALL     QPS          DIST COMPS
LuceneExact()                     1.000        6.722      1183514
LuceneCluster(n_probes=5)         0.680      618.438         6876
LuceneCluster(n_probes=10)        0.766      335.956        12456
LuceneCluster(n_probes=20)        0.835      173.782        23286
LuceneCluster(n_probes=50)        0.905       72.747        54539
LuceneCluster(n_probes=100)       0.948       37.339       105303
```

To me this looks a lot better and is in line with the speed-up we saw in the ES coarse quantization prototype.

**Pull request closed without merging**